src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,304 @@
+/*
+   +----------------------------------------------------------------------+
+   | Xdebug                                                               |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2002-2021 Derick Rethans                               |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 1.01 of the Xdebug license,   |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available at through the world-wide-web at                           |
+   | https://xdebug.org/license.php                                       |
+   | If you did not receive a copy of the Xdebug license and are unable   |
+   | to obtain it through the world-wide-web, please send a note to       |
+   | derick@xdebug.org so we can mail you a copy immediately.             |
+   +----------------------------------------------------------------------+
+ */
+#include "php.h"
+#include "ext/standard/php_string.h"
+
+#include "php_xdebug.h"
+#include "tracing_private.h"
+#include "trace_flamegraph.h"
+
+#include "lib/lib_private.h"
+#include "lib/var_export_line.h"
+
+extern ZEND_DECLARE_MODULE_GLOBALS(xdebug);
+
+static flamegraph_function *fg_function_ctor()
+{
+	flamegraph_function *ret;
+	ret = xdmalloc(sizeof(flamegraph_function));
+	ret->value = 0;
+	ret->prefix = NULL;
+	return ret;
+}
+
+static inline void fg_function_dtor(flamegraph_function *function)
+{
+	if (NULL != function->prefix) {
+		xdfree(function->prefix);
+	}
+	xdfree(function);
+}
+
+static inline xdebug_str *fg_function_key(const int function_nr) {
+	xdebug_str *key = xdebug_str_new();
+
+	xdebug_str_add_fmt(key, "fun-%d", function_nr);
```

Comment:
Please note that I added the "fun-" prefix only to see if would stop `xdebug_hash_add()` segfaulting.